### ISSUE TITLE
SSL connection error

### ISSUE THREAD

**Issue by Catastropha** [PyMySQL/mysqlclient-python#428]
Getting this error when trying to apply migrations with `python manage.py migrate`:
`django.db.utils.OperationalError: (2026, 'SSL connection error: error:1425F102:SSL routines:ssl_choose_client_version:unsupported protocol')`

Using Django 3.0.5 with python3.8 with mysqlclient==1.4.6 on ubuntu 20.04
These are the settings:
```
DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.mysql',
        'NAME': 'name',
        'USER': 'user',
        'PASSWORD': 'pass',
        'HOST': 'host',
        'PORT': 3306,
    }
}
```

Though, it is successfully connecting to the database using mysql-client without SSL

**Comment by kellycampbell** [comment]
I'm running into this as well. It seems to be because mysql 5.6 only supports TLSv1.0, and Ubuntu 20.04's openssl version defaults to requiring minimum of TLSv1.2.

https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/CHAP_MySQL.html#MySQL.Concepts.SSLSupport

There's a possible code workaround at the end of this bug thread that could be applied to the mysql client C code:
https://bugs.launchpad.net/ubuntu/+source/mysql-8.0/+bug/1872541

**Comment by kellycampbell** [comment]
Found the solution here:
https://askubuntu.com/questions/1233186/ubuntu-20-04-how-to-set-lower-ssl-security-level/1233456#1233456?newreg=0ce99c5c24274078a3f4c574fe68b52b

**Comment by ciarancourtney** [comment]
> Found the solution here:
> https://askubuntu.com/questions/1233186/ubuntu-20-04-how-to-set-lower-ssl-security-level/1233456#1233456?newreg=0ce99c5c24274078a3f4c574fe68b52b

Could you describe fully how you did so?

**Comment by AcckiyGerman** [comment]
I confirm the same bug.

## workarounds does not work
### 1. The [solution from askubuntu](https://askubuntu.com/questions/1233186/ubuntu-20-04-how-to-set-lower-ssl-security-level/1233456)
 did not help me (it may be I do something wrong).

However, even if the solution by lowering the ssl security would help:
- it is a bit tricky to implement and user have to be a root
- it can negatively affect the whole Ubuntu 20 system security
- implementing differs in different OS

### 2. [bug discussion](https://bugs.launchpad.net/ubuntu/+source/mysql-8.0/+bug/1872541) from the launchpad.net
the mysql server does not demand the SSL connection from the client, but only offers.
So one can avoid the bug
- by passing `--ssl-mode=DISABLED` to the mysql client
- by adding `ssl_mode = DISABLED` to the `[client]` section of mysql config file (e.g. `my.cnf`)

This solution works for the `mysql-client` application, but unfortunately, does not for `mysqlclient-python` library.

I've set up my django like this:
```python
'OPTIONS': {
      'read_default_file': os.path.join(BASE_DIR, 'mysql.cnf')
}
```
and created `mysql.cnf` file:
```
[client]
database = database
host = 127.0.0.1
user = myuser
password = mypassword
ssl_mode = DISABLED
```
Despite all other options (like host, user, password) was used, the lib skip the `ssl_mode` option and I've got the same error
```
django.db.utils.OperationalError: (2026, 'SSL connection error: error:1425F102:SSL routines:ssl_choose_client_version:unsupported protocol')
```

---
I think the bug has to be solved by following the second way (use `ssl_mod` option from the options file) or by adding a appropriate option to the connection settings.

**Comment by cprieto** [comment]
I was facing this issue a few days ago and after a while I found a _workaround_ to make it work, at least in my case it affects stand alone installations so we have the freedom to try a few things, what worked _at the end_ was in the mentioned bug report https://bugs.launchpad.net/ubuntu/+source/mysql-8.0/+bug/1872541 with this line:

 > Alternative solution: Using instead mariadb-client works without problems.

So this is what we did:

 - Uninstalled `mysqlclient`
 - Uninstalled `libmysqlclient-dev`
 - Installed `libmariadb-dev`
 - Made a soft link between `mariadb_config` as `mysql_config` (`ln -s /usr/bin/mariadb_config /usr/bin/mysql_config`)
 - Installed `mysqlclient` again

After this the connector worked without problems, standalone or using `mysql+mysqldb` in SQLAlchemy.

I don't think we can do much until something in `libmysql` changes (at least with latest versions of Ubuntu)

**Comment by ciarancourtney** [comment]
> I confirm the same bug.
> 
> ## Workarounds that didn't work for me
> ### 1. The [solution from askubuntu](https://askubuntu.com/questions/1233186/ubuntu-20-04-how-to-set-lower-ssl-security-level/1233456)
> did not help me (it may be I do something wrong).
> 
> However, even if the solution by lowering the ssl security would help:
> 
> * it is a bit tricky to implement and user have to be a root
> * it can negatively affect the whole Ubuntu 20 system security
> * implementing differs in different OS
> 
> ### 2. [bug discussion](https://bugs.launchpad.net/ubuntu/+source/mysql-8.0/+bug/1872541) from the launchpad.net
> the mysql server does not demand the SSL connection from the client, but only offers.
> So one can avoid the bug
> 
> * by passing `--ssl-mode=DISABLED` to the mysql client
> * by adding `ssl_mode = DISABLED` to the `[client]` section of mysql config file (e.g. `my.cnf`)
> 
> This solution works for the `mysql-client` application, but unfortunately, does not for `mysqlclient-python` library.
> 
> I've set up my django like this:
> 
> ```python
> 'OPTIONS': {
>       'read_default_file': os.path.join(BASE_DIR, 'mysql.cnf')
> }
> ```
> 
> and created `mysql.cnf` file:
> 
> ```
> [client]
> database = database
> host = 127.0.0.1
> user = myuser
> password = mypassword
> ssl_mode = DISABLED
> ```
> 
> Despite all other options (like host, user, password) was used, the lib skip the `ssl_mode` option and I've got the same error
> 
> ```
> django.db.utils.OperationalError: (2026, 'SSL connection error: error:1425F102:SSL routines:ssl_choose_client_version:unsupported protocol')
> ```
> 
> I think the bug has to be solved by following the second way (use `ssl_mod` option from the options file) or by adding a appropriate option to the connection settings.
> 
> ## P.S. it could be the Django bug
> I think so, because `manage.py dbshell` works for me, while `manage.py runserver` and `manage.py test` does not.
> I'm not so good programmer to find, where is the difference :(

ATM I workaround this by installing the latest pkg from github: 

```bash
pip install git+git://github.com/PyMySQL/mysqlclient-python.git@4c1950148ba10e23017bd5190f9da6bdb347e6b3
```

And then add the Django setting:

```python
    'OPTIONS' = {
        'ssl_mode': 'DISABLED',
    }
```

@methane are there any blockers to release master as v2 or v1.5? I've been using the commit above for months now

**Comment by methane** [comment]
I released v2.0.0 now.
https://pypi.org/project/mysqlclient/

**Comment by vhermecz** [comment]
If calling `MySQLdb.connect` function directly, using `MySQLdb.connect(..., ssl_mode="DISABLED")` solves the problem.
